internal/version/service.go

```diff
@@ -0,0 +1,163 @@
+		config:  config,
+		info:    info,
+		service: service,
+		Loop:    sync2.NewCycle(config.CheckInterval),
+		checked: make(chan struct{}, 0),
+		allowed: false,
+	}
+}
+
+// Run logs the current version information
+func (srv *Service) Run(ctx context.Context) error {
+	firstCheck := true
+	return srv.Loop.Run(ctx, func(ctx context.Context) error {
+		var err error
+		allowed, err := srv.checkVersion(ctx)
+
+		srv.mu.Lock()
+		srv.allowed = allowed
+		srv.mu.Unlock()
+
+		if firstCheck {
+			close(srv.checked)
+			firstCheck = false
+			if !allowed {
+				zap.S().Fatal(errOldVersion)
+			}
+		}
+
+		if err != nil {
+			zap.S().Errorf("Failed to do periodic version check: ", err)
+		}
+
+		return nil
+	})
+}
+
+// IsUpToDate returns whether if the Service is allowed to operate or not
+func (srv *Service) IsUpToDate() bool {
+	<-srv.checked
+
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	return srv.allowed
+}
+
+// CheckVersion checks if the client is running latest/allowed code
+func (srv *Service) checkVersion(ctx context.Context) (allowed bool, err error) {
+	defer mon.Task()(&ctx)(&err)
+	accepted, err := srv.queryVersionFromControlServer(ctx)
+	if err != nil {
+		return true, err
+	}
+
+	zap.S().Debugf("allowed versions from Control Server: %v", accepted)
+
+	list := getFieldString(&accepted, srv.service)
+	if list == nil {
+		return true, errs.New("Empty List from Versioning Server")
+	}
+	if containsVersion(list, srv.info.Version) {
+		zap.S().Infof("running on version %s", srv.info.Version.String())
+		allowed = true
+	} else {
+		zap.S().Errorf("running on not allowed/outdated version %s", srv.info.Version.String())
+		allowed = false
+	}
+	return
+}
+
+// QueryVersionFromControlServer handles the HTTP request to gather the allowed and latest version information
+func (srv *Service) queryVersionFromControlServer(ctx context.Context) (ver AllowedVersions, err error) {
+	// Tune Client to have a custom Timeout (reduces hanging software)
+	client := http.Client{
+		Timeout: srv.config.RequestTimeout,
+	}
+
+	// New Request that used the passed in context
+	req, err := http.NewRequest("GET", srv.config.ServerAddress, nil)
+	if err != nil {
+		return AllowedVersions{}, err
+	}
+	req = req.WithContext(ctx)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return AllowedVersions{}, err
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+
+	err = json.NewDecoder(resp.Body).Decode(&ver)
+	return
+}
+
+// DebugHandler returns a json representation of the current version information for the binary
+func (srv *Service) DebugHandler(w http.ResponseWriter, r *http.Request) {
+	j, err := Build.Marshal()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+
+	_, err = w.Write(j)
+	if err != nil {
+		zap.S().Errorf("error writing data to client %v", err)
+	}
+}
+
+func getFieldString(array *AllowedVersions, field string) []SemVer {
```

Comment:
using reflect is very eyebrow raising. okay for now but definitely prefer another way

Reply:
Do you have a better way to take the correct struct list?
I did not found a good solution for that and a map instead of the struct feels bad.

I thought about adding a function to the struct somehow

Reply:
i think parsing the json as a map is the most expected way tbh

Reply:
Not sure how to go about this now. Its it fine to leave and improve later?

Reply:
okay